### PR TITLE
Support backward compatibility cloudstack datacenter config

### PR DESCRIPTION
**Pull request by wongni**
*Issue #, if available:*

*Description of changes:*
Cloudstack provider has datacenter config but it contains the original values before transforming using Defaulters. This PR uses the passed-in clusterSpec instead of the stored datacenter config when preflight checking and template generation.

*Testing (if applicable):*

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: pkg/providers/cloudstack/cloudstack.go [hunk 1]
```diff
@@ -279,7 +279,22 @@ func (p *cloudstackProvider) UpdateKubeConfig(_ *[]byte, _ string) error {
 }
 
 func (p *cloudstackProvider) BootstrapClusterOpts() ([]bootstrapper.BootstrapClusterOption, error) {
-	return common.BootstrapClusterOpts(p.datacenterConfig.Spec.ManagementApiEndpoint, p.clusterConfig)
+	env := map[string]string{}
```
Comment (jiayiwang7) [inline pkg/providers/cloudstack/cloudstack.go:282]: I would keep using the original `common.BootstrapClusterOpts` and then add cloudstack custom NO_PROXY here to avoid duplicate code.

I'd also add clusterSpec as input arg, so that you can use clusterSpec.CloudStackDatacenter directly and remove https://github.com/aws/eks-anywhere/blob/18662f1c337379f49753a0e377b6dcdd83fa73d9/pkg/providers/cloudstack/cloudstack.go#L259 completely from the struct.

It is ok to change the interface method. For any other provider, just do
`func (p *vsphereProvider) BootstrapClusterOpts(_ *cluster.Spec) ([]bootstrapper.BootstrapClusterOption, error)`

File: pkg/providers/cloudstack/cloudstack.go [hunk 2]
```diff
@@ -166,7 +167,7 @@ func (p *cloudstackProvider) ValidateNewSpec(ctx context.Context, cluster *types
 		return err
 	}
 
-	datacenter := p.datacenterConfig
+	datacenter := clusterSpec.CloudStackDatacenter
```
Comment (jiayiwang7) [inline pkg/providers/cloudstack/cloudstack.go:170]: can we remove this? seems a bit redundant.

File: pkg/providers/cloudstack/cloudstack.go [hunk 3]
```diff
@@ -166,7 +167,7 @@ func (p *cloudstackProvider) ValidateNewSpec(ctx context.Context, cluster *types
...
 
 	oSpec := prevDatacenter.Spec
 	nSpec := datacenter.Spec
```
Comment (jiayiwang7) [inline pkg/providers/cloudstack/cloudstack.go:173]:
```suggestion
	nSpec := clusterSpec.CloudStackDatacenter.Spec
```

File: pkg/providers/cloudstack/cloudstack.go [hunk 4]
```diff
@@ -429,20 +444,20 @@ func (p *cloudstackProvider) validateSecretsUnchanged(ctx context.Context, clust
 }
 
 func secretDifferentFromProfile(secret *corev1.Secret, profile decoder.CloudStackProfileConfig) bool {
-	return string(secret.Data["uri"]) != profile.ManagementUrl ||
-		string(secret.Data["apikey"]) != profile.ApiKey ||
-		string(secret.Data["secretkey"]) != profile.SecretKey ||
-		string(secret.Data["verifyssl"]) != profile.VerifySsl
+	return string(secret.Data["api-url"]) != profile.ManagementUrl ||
```
Comment (jiayiwang7) [inline pkg/providers/cloudstack/cloudstack.go:447]: Why are we doing this? Seems unrelated to this PR

File: pkg/providers/cloudstack/cloudstack.go [hunk 5]
```diff
@@ -641,10 +654,7 @@ func (cs *CloudStackTemplateBuilder) GenerateCAPISpecControlPlane(clusterSpec *c
 	if clusterSpec.Cluster.Spec.ExternalEtcdConfiguration != nil {
 		etcdMachineSpec = *cs.etcdMachineSpec
 	}
-	if err != nil {
-		return nil, fmt.Errorf("failed to parse environment variable exec config: %v", err)
-	}
-	values := buildTemplateMapCP(clusterSpec, *cs.datacenterConfigSpec, *cs.controlPlaneMachineSpec, etcdMachineSpec)
+	values := buildTemplateMapCP(clusterSpec, clusterSpec.CloudStackDatacenter.Spec, *cs.controlPlaneMachineSpec, etcdMachineSpec)
```
Comment (jiayiwang7) [inline pkg/providers/cloudstack/cloudstack.go:657]: I think you can remove the second arg since it can be fetched from first arg.

```suggestion
	values := buildTemplateMapCP(clusterSpec, *cs.controlPlaneMachineSpec, etcdMachineSpec)
```

File: pkg/providers/cloudstack/cloudstack.go [hunk 6]
```diff
@@ -661,7 +671,7 @@ func (cs *CloudStackTemplateBuilder) GenerateCAPISpecControlPlane(clusterSpec *c
 func (cs *CloudStackTemplateBuilder) GenerateCAPISpecWorkers(clusterSpec *cluster.Spec, workloadTemplateNames, kubeadmconfigTemplateNames map[string]string) (content []byte, err error) {
 	workerSpecs := make([][]byte, 0, len(clusterSpec.Cluster.Spec.WorkerNodeGroupConfigurations))
 	for _, workerNodeGroupConfiguration := range clusterSpec.Cluster.Spec.WorkerNodeGroupConfigurations {
-		values := buildTemplateMapMD(clusterSpec, *cs.datacenterConfigSpec, cs.WorkerNodeGroupMachineSpecs[workerNodeGroupConfiguration.MachineGroupRef.Name], workerNodeGroupConfiguration)
+		values := buildTemplateMapMD(clusterSpec, clusterSpec.CloudStackDatacenter.Spec, cs.WorkerNodeGroupMachineSpecs[workerNodeGroupConfiguration.MachineGroupRef.Name], workerNodeGroupConfiguration)
```
Comment (jiayiwang7) [inline pkg/providers/cloudstack/cloudstack.go:674]: Same as above, remove second arg

File: pkg/providers/cloudstack/cloudstack.go [hunk 7]
```diff
@@ -769,7 +779,7 @@ func buildTemplateMapCP(clusterSpec *cluster.Spec, datacenterConfigSpec v1alpha1
 	}
 
 	if clusterSpec.Cluster.Spec.ProxyConfiguration != nil {
-		fillProxyConfigurations(values, clusterSpec, datacenterConfigSpec)
+		fillProxyConfigurations(values, clusterSpec, clusterSpec.CloudStackDatacenter.Spec)
```
Comment (jiayiwang7) [inline pkg/providers/cloudstack/cloudstack.go:782]: Same, remove third arg.
```suggestion
		fillProxyConfigurations(values, clusterSpec)
```

File: pkg/providers/cloudstack/cloudstack.go [hunk 8]
```diff
@@ -906,7 +916,7 @@ func (p *cloudstackProvider) getControlPlaneNameForCAPISpecUpgrade(ctx context.C
 	if err != nil {
 		return "", err
 	}
-	needsNewControlPlaneTemplate := needsNewControlPlaneTemplate(currentSpec, newClusterSpec, csdc, p.datacenterConfig, controlPlaneVmc, controlPlaneMachineConfig)
+	needsNewControlPlaneTemplate := needsNewControlPlaneTemplate(currentSpec, newClusterSpec, csdc, newClusterSpec.CloudStackDatacenter, controlPlaneVmc, controlPlaneMachineConfig)
```
Comment (jiayiwang7) [inline pkg/providers/cloudstack/cloudstack.go:919]: Same, remove `newClusterSpec.CloudStackDatacenter`

File: pkg/providers/cloudstack/cloudstack.go [hunk 9]
```diff
@@ -969,7 +979,7 @@ func (p *cloudstackProvider) getEtcdTemplateNameForCAPISpecUpgrade(ctx context.C
 	if err != nil {
 		return "", err
 	}
-	needsNewEtcdTemplate := needsNewEtcdTemplate(currentSpec, newClusterSpec, csdc, p.datacenterConfig, etcdMachineVmc, etcdMachineConfig)
+	needsNewEtcdTemplate := needsNewEtcdTemplate(currentSpec, newClusterSpec, csdc, newClusterSpec.CloudStackDatacenter, etcdMachineVmc, etcdMachineConfig)
```
Comment (jiayiwang7) [inline pkg/providers/cloudstack/cloudstack.go:982]: Remove `newClusterSpec.CloudStackDatacenter`

File: pkg/providers/cloudstack/cloudstack.go (hunk 4 above)
Comment (wongni) [reply pkg/providers/cloudstack/cloudstack.go:436]: I found this issue (secret data name) when testing upgrade test.

File: pkg/workflows/create_test.go [hunk 10]
```diff
@@ -86,7 +86,7 @@ func (c *createTestSetup) expectCreateBootstrap() {
 	}
 
 	gomock.InOrder(
-		c.provider.EXPECT().BootstrapClusterOpts().Return(opts, nil),
+		c.provider.EXPECT().BootstrapClusterOpts(gomock.Any()).Return(opts, nil),
```
Comment (jiayiwang7) [inline pkg/workflows/create_test.go:89]: can we be specific and use clusterSpec?

File: pkg/workflows/delete_test.go [hunk 11]
```diff
@@ -61,7 +61,7 @@ func (c *deleteTestSetup) expectCreateBootstrap() {
 	}
 
 	gomock.InOrder(
-		c.provider.EXPECT().BootstrapClusterOpts().Return(opts, nil),
+		c.provider.EXPECT().BootstrapClusterOpts(gomock.Any()).Return(opts, nil),
```
Comment (jiayiwang7) [inline pkg/workflows/delete_test.go:64]: can we be specific and use clusterSpec?

File: pkg/workflows/delete_test.go [hunk 12]
```diff
@@ -76,7 +76,7 @@ func (c *deleteTestSetup) expectNotToCreateBootstrap() {
 		bootstrapper.WithDefaultCNIDisabled(), bootstrapper.WithExtraDockerMounts(),
 	}
 
-	c.provider.EXPECT().BootstrapClusterOpts().Return(opts, nil).Times(0)
+	c.provider.EXPECT().BootstrapClusterOpts(gomock.Any()).Return(opts, nil).Times(0)
```
Comment (jiayiwang7) [inline pkg/workflows/delete_test.go:79]: can we be specific and use clusterSpec?

File: pkg/workflows/upgrade_test.go [hunk 13]
```diff
@@ -186,7 +186,7 @@ func (c *upgradeTestSetup) expectCreateBootstrap() {
 	}
 
 	gomock.InOrder(
-		c.provider.EXPECT().BootstrapClusterOpts().Return(opts, nil),
+		c.provider.EXPECT().BootstrapClusterOpts(gomock.Any()).Return(opts, nil),
```
Comment (jiayiwang7) [inline pkg/workflows/upgrade_test.go:189]: can we be specific and use clusterSpec?

File: pkg/workflows/upgrade_test.go [hunk 14]
```diff
@@ -198,7 +198,7 @@ func (c *upgradeTestSetup) expectCreateBootstrap() {
 }
 
 func (c *upgradeTestSetup) expectNotToCreateBootstrap() {
-	c.provider.EXPECT().BootstrapClusterOpts().Times(0)
+	c.provider.EXPECT().BootstrapClusterOpts(gomock.Any()).Times(0)
```
Comment (jiayiwang7) [inline pkg/workflows/upgrade_test.go:201]: can we be specific and use clusterSpec?

File: pkg/workflows/upgrade_test.go [hunk 15]
```diff
@@ -389,7 +389,7 @@ func (c *upgradeTestSetup) expectPauseGitOpsKustomizationNotToBeCalled() {
 }
 
 func (c *upgradeTestSetup) expectCreateBootstrapNotToBeCalled() {
-	c.provider.EXPECT().BootstrapClusterOpts().Times(0)
+	c.provider.EXPECT().BootstrapClusterOpts(gomock.Any()).Times(0)
```
Comment (jiayiwang7) [inline pkg/workflows/upgrade_test.go:392]: can we be specific and use clusterSpec?

File: pkg/providers/common/common.go [hunk 16]
```diff
@@ -30,18 +30,19 @@ func GetAuditPolicy() string {
 	return auditPolicy
 }
 
-func BootstrapClusterOpts(serverEndpoint string, clusterConfig *v1alpha1.Cluster) ([]bootstrapper.BootstrapClusterOption, error) {
+func BootstrapClusterOpts(clusterConfig *v1alpha1.Cluster, serverEndpoints ...string) ([]bootstrapper.BootstrapClusterOption, error) {
```
Comment (jiayiwang7) [inline pkg/providers/common/common.go:33]: Does it make sense for all providers? Having a serverEndpoints that will be added to NoProxy. Or is this specific to cloudstack? If so we might not wanna update this common method

Comment (wongni) [reply pkg/providers/common/common.go:33]: Currently, only cloudstack and vsphere use common.BootstrapClusterOpts and they both want to add server endpoint(s) to NoProxy. I think this makes sense for all providers.

**Review by jiayiwang7 — APPROVED** [review]
